ifm-web: init at 4.0.2

litchipi opened the pull request:
## Description of changes

Context: https://github.com/NixOS/nixpkgs/pull/328205#pullrequestreview-2236375650

Create a proper package for `ifm`, with the name `ifm-web` as ifm was already taken.
Allows for more meta informations, as well as licensing

Also creates a serve script, that allowed to greatly improve the module serving script.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

litchipi committed:
ifm-web: init at 4.0.2

Signed-off-by: Litchi Pi <litchi.pi@proton.me>

SuperSandro2000 commented on pkgs/by-name/if/ifm-web/default.nix:
```diff
@@ -0,0 +1,48 @@
+  phases = "installPhase";
+  installPhase = ''
+    mkdir -p $out/bin $out/php
+    ln -s $src $out/php/index.php
```
We should copy the files to not retain a source reference and allow changes to the files.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5436,7 +5436,7 @@ with pkgs;
 
   ifm = callPackage ../tools/graphics/ifm { };
 
-  ifm-web = callPackage ../by-name/if/ifm-web/package.nix { };
```
This line is not required and is done by nixpkgs code already.

SuperSandro2000 reviewed (dismissed):
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) when naming your commits.

litchipi commented on pkgs/by-name/if/ifm-web/default.nix:
```diff
@@ -0,0 +1,48 @@
+  pname = "ifm";
+  inherit version src;
+
+  phases = "installPhase";
```
I'm affraid that's something I do regularly on my own nix derivations, didn't know it was a problem

SuperSandro2000 replied:
It disables things like patches or fixupPhase

litchipi commented:
Fixed and squashed to a clean commit

@SuperSandro2000 Is the naming of my commit alright now ? I tried to follow the contributing guidelines.

That would be cool to have a CI for this tho, not the first time I mess up the commit messages

pbsds commented on pkgs/by-name/if/ifm-web/package.nix:
```diff
@@ -0,0 +1,46 @@
+  pname = "ifm";
+  inherit version src;
+
+  installPhase = ''
+    mkdir -p $out/bin $out/php
+    cp $src $out/php/index.php
+    cp ${serve_script} $out/bin/ifm
+  '';
```
```suggestion
  installPhase = ''
    runHook preInstall
    mkdir -p $out/bin $out/php
    cp $src $out/php/index.php
    cp ${serve_script} $out/bin/ifm
    runHook postInstall
  '';
```

pbsds commented on nixos/modules/services/web-apps/ifm.nix:
```diff
@@ -55,24 +53,20 @@ in {
         StandardOutput = "journal";
         BindPaths = "${cfg.dataDir}:/data";
         PrivateTmp = true;
+        ExecStart = "${lib.getExe pkgs.ifm-web} ${cfg.listenAddress} ${builtins.toString cfg.port} /data";
```
```suggestion
        ExecStart = "${lib.getExe pkgs.ifm-web} ${lib.escapeShellArg cfg.listenAddress} ${builtins.toString cfg.port} /data";
```

gador commented on pkgs/by-name/if/ifm-web/package.nix:
```diff
@@ -0,0 +1,48 @@
+    runHook postInstall
+  '';
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

gador commented on pkgs/by-name/if/ifm-web/package.nix:
```diff
@@ -0,0 +1,48 @@
+    '';
+    homepage = "https://github.com/misterunknown/ifm";
+    changelog = "https://github.com/misterunknown/ifm/releases/tag/v${version}";
+    license = licenses.mit;
```
```suggestion
    license = lib.licenses.mit;
```

gador commented on pkgs/by-name/if/ifm-web/package.nix:
```diff
@@ -0,0 +1,48 @@
+    homepage = "https://github.com/misterunknown/ifm";
+    changelog = "https://github.com/misterunknown/ifm/releases/tag/v${version}";
+    license = licenses.mit;
+    maintainers = with maintainers; [ litchipi ];
```
```suggestion
    maintainers = with lib.maintainers; [ litchipi ];
```

gador reviewed (dismissed):
Small nitpicks, otherwise looks good!

pbsds commented on nixos/modules/services/web-apps/ifm.nix:
```diff
@@ -61,18 +59,13 @@ in {
         IFM_ROOT_DIR = "/data";
```
```suggestion
```

pbsds commented on pkgs/by-name/if/ifm-web/package.nix:
```diff
@@ -0,0 +1,50 @@
+    cp ${serve_script} $out/bin/ifm
+    runHook postInstall
+  '';
+
```
```suggestion

  passthru.tests = {
    inherit (nixosTests) ifm;
  };

```

pbsds commented on pkgs/by-name/if/ifm-web/package.nix:
```diff
@@ -0,0 +1,50 @@
+  stdenv,
+  lib,
+  php83,
+  writeShellScript,
```
```suggestion
  writeShellScript,
  nixosTests,
```

pbsds commented:
in reference to https://github.com/NixOS/nixpkgs/pull/334559#pullrequestreview-2237574618, i believe `ifm-web: init at 4.0.2` is appropriate

litchipi commented:
Fixed, squashed and rebased.

I also allowed myself to fix the release notes (the "Available as services.ifm" was missing)

gador commented:
failing test for the manual:
```
bad local reference, id #opt-services.ifm not known
```

pbsds commented:
thank you!
